README.md

```diff
@@ -1,16 +1,1261 @@
+
+### 3.1.2 Ability
+
+The `do` field MUST contain a [UCAN Ability](https://github.com/ucan-wg/spec/#23-ability). This field can be thought of as the message or trait being sent to the resource.
+
+### 3.1.3 Inputs
+
+The `inputs` field MUST contain any arguments expected by the URI/Ability pair. This MAY be different between different URIs and Abilities, and is thus left to the executor to define the shape of this data.
+
+### 3.2 DAG-JSON Examples
+
+Interacting with an HTTP API:
+
+``` json
+{
+  "with": "https://example.com/blog/posts",
+  "do": "crud/create",
+  "inputs": {
+    "headers": {
+      "content-type": "application/json"
+    },
+    "payload": {
+      "title": "How UCAN Tasks Changed My Life",
+      "body": "This is the story of how one spec changed everything...",
+      "topics": ["authz", "journal"],
+      "draft": true
+    }
+  }
+}
+```
+
+Sending Email:
+
+``` json
+{
+  "with": "mailto:akiko@example.com",
+  "do": "msg/send",
+  "inputs": {
+    "to": ["boris@example.com", "carol@example.com"],
+    "subject": "Coffee",
+    "body": "Hey you two, I'd love to get coffee sometime and talk about UCAN Tasks!"
+  }
+}
+```
+
+Running WebAssembly from binary:
+
+``` json
+{
+  "with": "data:application/wasm;base64,AHdhc21lci11bml2ZXJzYWwAAAAAAOAEAAAAAAAAAAD9e7+p/QMAkSAEABH9e8GowANf1uz///8UAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAP////8AAAAACAAAACoAAAAIAAAABAAAACsAAAAMAAAACAAAANz///8AAAAA1P///wMAAAAlAAAALAAAAAAAAAAUAAAA/Xu/qf0DAJHzDx/44wMBqvMDAqphAkC5YAA/1mACALnzB0H4/XvBqMADX9bU////LAAAAAAAAAAAAAAAAAAAAAAAAAAvVXNlcnMvZXhwZWRlL0Rlc2t0b3AvdGVzdC53YXQAAGFkZF9vbmUHAAAAAAAAAAAAAAAAYWRkX29uZV9mAAAADAAAAAAAAAABAAAAAAAAAAkAAADk////AAAAAPz///8BAAAA9f///wEAAAAAAAAAAQAAAB4AAACM////pP///wAAAACc////AQAAAAAAAAAAAAAAnP///wAAAAAAAAAAlP7//wAAAACM/v//iP///wAAAAABAAAAiP///6D///8BAAAAqP///wEAAACk////AAAAAJz///8AAAAAlP///wAAAACM////AAAAAIT///8AAAAAAAAAAAAAAAAAAAAAAAAAAET+//8BAAAAWP7//wEAAABY/v//AQAAAID+//8BAAAAxP7//wEAAADU/v//AAAAAMz+//8AAAAAxP7//wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAU////pP///wAAAAAAAQEBAQAAAAAAAACQ////AAAAAIj///8AAAAAAAAAAAAAAADQAQAAAAAAAA==",
+  "do": "wasm/run",
+  "inputs": {
+    "func": "add_one",
+    "args": [42]
+  }
+}
+```
+
+Executing all of the capabilities in a UCAN directly:
+
+``` json
+{
+  "with": "ipfs://bafkreiemaanh3kxqchhcdx3yckeb3xvmboztptlgtmnu5jp63bvymxtlva",
+  "do": "ucan/run",
+  "inputs": null
+}
+```
+
+# 4 Task
+
+A Task is subtype of a [Closure](#3-closure), adding an OPTIONAL metadata field. If not present, the `meta` field defaults to an empty map. A Task can be trivially converted to a Closure by removing the `meta` field.
+
+``` ipldsch
+type Task struct {
+  with   URI
+  do     Ability
+  inputs Any
+  meta {String : Any} (implicit {})
```

Comment:
Why does `meta` needs to be special as opposed be part of `input` for example ? In our system that is where we stick it, perhaps because we just have `nb` but I think it's question worth addressing.

Reply:
I'm guessing the distinction here is contents of meta are expected not to change the result of the computation.

F.ex: if I have an input that's some source code, and some "meta" that says "please give me two CPUs", one of those *should* impact the outcome, and one should not (beyond at most, perhaps causing failure or failure to schedule, which are distinguishable situations).

This would be relevant to memoization.

Reply:
Yes, exactly this. The principle is that for memoization of job results, `with`, `inputs` and `do` matter, but `meta` does not. Or I describe it as the other fields are "what we are executing" and anything in `meta` is "how we are executing it".

Reply:
> Or I describe it as the other fields are "what we are executing" and anything in meta is "how we are executing it".

Or another way: "anything that doesn't get memoized". User writable tags, scheduler hints, cache settings, etc. @zeeshanlakhani has suggested that we break it out into multiple optional fields, at least for IPVM things: `metadata`, `resources`, etc

In UCAN core we had a lot of success with open fields that let people extend the system. We want _something_ like that, but what if we call that field `meta` or `ext` or broken into several fields is totally open.

Reply:
All these makes sense & probably worth capturing in the spec as well so implementers are aware this.